pkg/registry/apis/provisioning/jobs/persistentstore.go

```diff
@@ -0,0 +1,466 @@
+	if !ok {
+		return apifmt.Errorf("unexpected object type %T", jobsObj)
+	}
+
+	for _, job := range jobs.Items {
+		if job.Labels == nil {
+			job.Labels = make(map[string]string)
+		}
+		delete(job.Labels, LabelJobClaim)
+		job.Status.State = provisioning.JobStatePending
+
+		// We list jobs from all namespaces. So when we want to update a specific job, we also need its namespace in the context.
+		ctx := request.WithNamespace(ctx, job.GetNamespace())
+		// Likewise, we should use the provisioning identity now that we have the namespace we are operating within.
+		ctx, _, err = identity.WithProvisioningIdentity(ctx, job.GetNamespace())
+		if err != nil {
+			// This should never happen, as it is already a valid namespace from the job existing... but better be safe.
+			return apifmt.Errorf("failed to get provisioning identity for '%s': %w", job.GetNamespace(), err)
+		}
+
+		timeoutCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		_, _, err := s.jobStore.Update(timeoutCtx,
+			job.GetName(),                       // name
+			rest.DefaultUpdatedObjectInfo(&job), // objInfo
+			failCreation,                        // createValidation
+			nil,                                 // updateValidation
+			false,                               // forceAllowCreate
+			&metav1.UpdateOptions{},             // options
+		)
+		cancel() // we have no response body to read, so just cancel immediately
+		if apierrors.IsConflict(err) || errors.Is(err, errWouldCreate) {
+			continue
+		}
+		if err != nil {
+			return apifmt.Errorf("failed to unclaim job '%s' in '%s': %w", job.GetName(), job.GetNamespace(), err)
+		}
+	}
+
+	return nil
+}
+
+// Insert adds a new job to the queue.
+//
+// The job name is not honoured. It will be overwritten with a name that fits the job.
+//
+// This saves it if it is a new job, or fails with `apierrors.IsAlreadyExists(err) == true` if one already exists.
+func (s *persistentStore) Insert(ctx context.Context, job *provisioning.Job) (*provisioning.Job, error) {
+	s.generateJobName(job) // Side-effect: updates the job's name.
+
+	obj, err := s.jobStore.Create(ctx, job, nil, &metav1.CreateOptions{})
+	if apierrors.IsAlreadyExists(err) {
+		return nil, apifmt.Errorf("job '%s' in '%s' already exists: %w", job.GetName(), job.GetNamespace(), err)
+	}
+	if err != nil {
+		return nil, apifmt.Errorf("failed to create job '%s' in '%s': %w", job.GetName(), job.GetNamespace(), err)
+	}
+
+	created, ok := obj.(*provisioning.Job)
+	if !ok {
+		return nil, apifmt.Errorf("unexpected object type %T", obj)
+	}
+
+	select {
+	case s.notifications <- struct{}{}:
+	default:
+		// We don't want to block if there is already a notification waiting.
+	}
+
+	return created, nil
+}
+
+func (s *persistentStore) InsertNotifications() chan struct{} {
+	return s.notifications
+}
+
+// generateJobName creates and updates the job's name to one that fits it.
+func (s *persistentStore) generateJobName(job *provisioning.Job) {
```

Comment:
I am still not too convinced about using name for this purpose and adding an additional HistoricalJobs.

Given that we don't have field selectors, did you evaluate the option to use purely labels? (you use already some for repository, claiming, etc). 

For example: 

`provisioning.grafana.app/repository` = `my-repo`
`provisioning.grafana.app/claim` = `1280218201100`
`provisioning.grafana.app/jobtype` = `sync`
`provisioning.grafana.app/historical` = true


And for admission maybe doing that on a job admission hook which prevents us from queueing what cannot be processed? 

In that admission hook we could also enforce other limits like the having no more than X jobs per namespace, etc.

Reply:
Admission hooks are a great idea! I just hadn't thought of them :)